bundles/org.openhab.binding.network/src/main/java/org/openhab/binding/network/internal/handler/SpeedTestHandler.java

```diff
@@ -0,0 +1,176 @@
+/**
+ * Copyright (c) 2010-2019 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.network.internal.handler;
+
+import static org.openhab.binding.network.internal.NetworkBindingConstants.*;
+
+import java.math.BigDecimal;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.core.library.types.DateTimeType;
+import org.eclipse.smarthome.core.library.types.DecimalType;
+import org.eclipse.smarthome.core.library.types.OnOffType;
+import org.eclipse.smarthome.core.library.types.QuantityType;
+import org.eclipse.smarthome.core.library.unit.SmartHomeUnits;
+import org.eclipse.smarthome.core.thing.ChannelUID;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.ThingStatusDetail;
+import org.eclipse.smarthome.core.thing.binding.BaseThingHandler;
+import org.eclipse.smarthome.core.types.Command;
+import org.eclipse.smarthome.core.types.RefreshType;
+import org.eclipse.smarthome.core.types.State;
+import org.eclipse.smarthome.core.types.UnDefType;
+import org.openhab.binding.network.internal.SpeedTestConfiguration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import fr.bmartel.speedtest.SpeedTestReport;
+import fr.bmartel.speedtest.SpeedTestSocket;
+import fr.bmartel.speedtest.inter.ISpeedTestListener;
+import fr.bmartel.speedtest.model.SpeedTestError;
+import fr.bmartel.speedtest.model.SpeedTestMode;
+
+/**
+ * The {@link SpeedTestHandler } is responsible for launching bandwidth
+ * measurements at a given interval and for given file / size
+ *
+ * @author Gaël L'hopital - Initial contribution
+ */
+@NonNullByDefault
+public class SpeedTestHandler extends BaseThingHandler implements ISpeedTestListener {
+    private final Logger logger = LoggerFactory.getLogger(SpeedTestHandler.class);
+    private static final BigDecimal TO_MBIT = new BigDecimal(1048576);
+    private @NonNullByDefault({}) SpeedTestSocket speedTestSocket;
+    private @NonNullByDefault({}) ScheduledFuture<?> refreshTask;
+    private @NonNullByDefault({}) SpeedTestConfiguration configuration;
+    private State bufferedProgress = UnDefType.UNDEF;
+
+    public SpeedTestHandler(Thing thing) {
+        super(thing);
+    }
+
+    @Override
+    public void initialize() {
+        speedTestSocket = new SpeedTestSocket(1500);
```

Comment:
I assume that speedtests are not executed that often. Wouldn't it be better to open and close the `speedTestSocket` only when it is needed (i.e. before l. 87)? You could also set it to `null` on completion and check if it is `null` before starting a new one (or log an error if the old one is still running).